src/session/components/avatar.rs

```diff
@@ -111,74 +124,127 @@ impl Avatar {
+                glib::Object,
+            >| {
+                if let Some(Ok(chat)) = item.map(|i| i.downcast::<Chat>()) {
+                    if chat.is_own_chat() {
+                        // Show bookmark icon for Saved Messages chat
+                        return Some("user-bookmarks-symbolic".to_string());
+                    }
+                }
+
+                None
+            }));
         icon_name_expression.bind(&*imp.avatar, "icon-name", Some(self));
-        custom_image_expression.bind(&*imp.avatar, "custom-image", Some(self));
-        show_initials_expression.bind(&*imp.avatar, "show-initials", Some(self));
-        text_expression.bind(&*imp.avatar, "text", Some(self));
+
+        // Don't show the initials if we show a custom icon
+        icon_name_expression
+            .chain_closure::<bool>(closure!(
+                |_: Self, icon_name: Option<String>| icon_name.is_none()
+            ))
+            .bind(&*imp.avatar, "show-initials", Some(self));
     }
 
-    fn request_avatar_image(&self) {
-        if let Some(item) = &*self.imp().item.borrow() {
-            item.set_needed(true);
+    fn load_image(&self, avatar_item: Option<AvatarItem>, session: Session) {
+        if let Some(avatar_item) = avatar_item {
+            let file = avatar_item.file;
+            if file.local.is_downloading_completed {
+                let texture = gdk::Texture::from_filename(&file.local.path).unwrap();
+                self.imp().avatar.set_custom_image(Some(&texture));
+            } else {
+                let (sender, receiver) =
+                    glib::MainContext::sync_channel::<File>(Default::default(), 5);
+
+                receiver.attach(
+                    None,
+                    clone!(@weak self as obj => @default-return glib::Continue(false), move |file| {
+                        if file.local.is_downloading_completed {
+                            let texture = gdk::Texture::from_filename(&file.local.path).unwrap();
+                            obj.imp().avatar.set_custom_image(Some(&texture));
+                        }
+                        glib::Continue(true)
+                    }),
+                );
+
+                session.download_file(file.id, sender);
+            }
+        } else {
+            self.imp().avatar.set_custom_image(gdk::Paintable::NONE);
         }
     }
 
-    pub(crate) fn item(&self) -> Option<AvatarItem> {
+    pub(crate) fn item(&self) -> Option<glib::Object> {
         self.imp().item.borrow().clone()
     }
 
-    pub(crate) fn set_item(&self, item: Option<AvatarItem>) {
-        self.imp().item.replace(item);
+    pub(crate) fn set_item(&self, item: Option<glib::Object>) {
+        let imp = self.imp();
 
-        self.request_avatar_image();
+        if let Some(old_item) = self.item() {
+            if let Some(id) = imp.handler_id.take() {
+                old_item.disconnect(id);
+            }
+        }
```

Comment:
I don't think it makes sense to test whether the old item is set. Just take the handler and disconnect it. 
```suggestion
        if let Some(id) = imp.handler_id.take() {
            old_item.disconnect(id);
        }
```

Reply:
Well, without the object I can't disconnect anything from it. Or did you mean to take the item directly instead?

Reply:
O sorry, I made the false suggestion. In fact, if the handler is `Some` some, the item cannot be `None`. 
```suggestion
        if let Some(id) = imp.handler_id.take() {
            self.item().unwrap().disconnect(id);
        }
```